scio-elasticsearch/es2/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java

```diff
@@ -259,11 +323,50 @@ public void processElement(ProcessContext c) throws Exception {
             Iterables.partition(actionRequests::iterator, maxBulkRequestSize);
 
         chunks.forEach(chunk -> {
+          Exception exception;
+
+          final BackOff backoff = backoffConfig.backoff();
+
+          do {
```

Comment:
@soundofjw wdyt of remove this `do { ... } while(true)` in favor of a 
```java
while (BackOffUtils.next(Sleeper.DEFAULT, backoff)) {
 ...
}
```

removes the need for all of this 

```java
try {
  final long sleepTime = backoff.nextBackOffMillis();
  if (sleepTime == BackOff.STOP) {
    break;
  }
  Thread.sleep(sleepTime);
} catch (InterruptedException | IOException e1) {
  LOG.error("Interrupt during backoff", e1);
  break;
}
```

Reply:
Your proposal will introduce a sleep at the beginning of every attempt:

```java
  public static boolean next(Sleeper sleeper, BackOff backOff)
      throws InterruptedException, IOException {
    long backOffTime = backOff.nextBackOffMillis();
    if (backOffTime == BackOff.STOP) {
      return false;
    }
    sleeper.sleep(backOffTime);
    return true;
  }```

Reply:
Yeah sorry for poor explanation, yeah correct... what I was trying to say is that I think we can refactor this block by decoupling the logic a little bit and at the same time trying to leverage that util for the retry logic ... I think we can get a better semantic and readability.